TypeError: Cannot read property 'name' of null

First off, thanks for the library! Overall it's been incredibly performant and straightforward to use. I am running into one issue which may be an issue with the library, though it's a bit hard to follow the protocol for me so my usage may be off as well.

I'm running into an uncaught error during the `parseComplete` handler in client.js (https://github.com/brianc/node-postgres/blob/v6.0.1/lib/client.js#L131).

I am doing a direct `connect` against the pool and then utilizing pg-query-stream

``` javascript
pool.connect().then(function(client){
  stream = client.query(new QueryStream(query, null));

  stream.on('error', function(streamError){
    stream.close(function(closeError){ 
      // What to do with a close error? At this point stream is already hosed :shrug:
      client.release(streamError);
    }
  });

  stream.on('close', function(){
    client.release();
  }); 
});
```

At some point during my app's life cycle, I get an uncaught error

```
TypeError: Cannot read property 'name' of null
    at null.<anonymous> (/var/app/current/node_modules/pg/lib/client.js:139:26)
    at emit (events.js:107:17)
    at Socket.<anonymous> (/var/app/current/node_modules/pg/lib/connection.js:121:12)
    at Socket.emit (events.js:107:17)
    at readableAddChunk (_stream_readable.js:163:16)
    at Socket.Readable.push (_stream_readable.js:126:10)
    at TCP.onread (net.js:540:20)
```

So obviously `activeQuery` is becoming null, and I am trying to narrow down why I would receive a `parseComplete` event when the active query is null.

According to my logs, the stream is _not_ emitting an error, and seems to be closing normally. Therefore it seems like the connection client is getting a `readyForQuery` or `end` event, and then right after that getting the `parseComplete` event. Any idea why this would be happening, or see any issues with my usage? Thanks for any help you can give! I'll keep looking into it as well.

I'm also using query_timeout:

[2019-07-11 09:36:36] [ERROR]  [pid:8718] TypeError: Cannot read property 'name' of null
TypeError: Cannot read property 'name' of null
    at Connection.<anonymous> (/opt/acn/4be1e1fb/api/node_modules/pg/lib/client.js:313:26)
    at emitOne (events.js:116:13)
    at Connection.emit (events.js:211:7)
    at Socket.<anonymous> (/opt/acn/4be1e1fb/api/node_modules/pg/lib/connection.js:126:12)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:597:20)

--- a/packages/pg/lib/query.js
+++ b/packages/pg/lib/query.js
@@ -31,7 +31,6 @@ class Query extends EventEmitter {
     this.isPreparedStatement = false
     this._canceledDueToError = false
     this._promise = null
-    this._hasSentSync = false
   }
 
   requiresPreparation() {

--- a/packages/pg/lib/query.js
+++ b/packages/pg/lib/query.js
@@ -101,19 +101,22 @@ class Query extends EventEmitter {
     this._checkForMultirow()
     this._result.addCommandComplete(msg)
     // need to sync after each command complete of a prepared statement
-    this.maybeSync(connection)
+    // if we were using a row count which results in multiple calls to _getRows
+    if (this.rows) {
+      this.maybeSync(connection)
+    }
   }
 
   // if a named prepared statement is created with empty query text
   // the backend will send an emptyQuery message but *not* a command complete message
   // execution on the connection will hang until the backend receives a sync message
   handleEmptyQuery(connection) {
-    this.maybeSync(connection)
+    // this.maybeSync(connection)
   }
 
   handleError(err, connection) {
     // need to sync after error during a prepared statement
-    this.maybeSync(connection)
+    // this.maybeSync(connection)
     if (this._canceledDueToError) {
       err = this._canceledDueToError
       this._canceledDueToError = false
@@ -143,7 +146,7 @@ class Query extends EventEmitter {
   // say "Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages:
   // CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended."
   maybeSync(connection) {
-    if (this.isPreparedStatement && !this._hasSentSync) {
+    if (this.isPreparedStatement) {
       this._hasSentSync = true
       connection.sync()
     }
@@ -181,7 +184,11 @@ class Query extends EventEmitter {
       portal: this.portal,
       rows: rows,
     })
-    connection.flush()
+    if (!rows) {
+      this.maybeSync(connection)
+    } else {
+      connection.flush()
+    }
   }
 
   // http://developer.postgresql.org/pgdocs/postgres/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY

--- a/packages/pg/lib/query.js
+++ b/packages/pg/lib/query.js
@@ -103,20 +103,22 @@ class Query extends EventEmitter {
     // need to sync after each command complete of a prepared statement
     // if we were using a row count which results in multiple calls to _getRows
     if (this.rows) {
-      this.maybeSync(connection)
+      connection.sync()
     }
   }
 
   // if a named prepared statement is created with empty query text
   // the backend will send an emptyQuery message but *not* a command complete message
-  // execution on the connection will hang until the backend receives a sync message
+  // since we pipeline sync immediately after execute we don't need to do anything here
+  // unless we have rows specified, in which case we did not pipeline the intial sync call
   handleEmptyQuery(connection) {
-    // this.maybeSync(connection)
+    if (this.rows) {
+      connection.sync()
+    }
   }
 
   handleError(err, connection) {
     // need to sync after error during a prepared statement
-    // this.maybeSync(connection)
     if (this._canceledDueToError) {
       err = this._canceledDueToError
       this._canceledDueToError = false
@@ -137,19 +139,6 @@ class Query extends EventEmitter {
       this.callback(null, this._results)
     }
     this.emit('end', this._results)
-  }
-
-  // In postgres 9.x & 10.x the backend sends both a CommandComplete and ErrorResponse
-  // to the same query when it times out due to a statement_timeout on rare, random occasions.  If we send sync twice we will receive
-  // to ReadyForQuery messages .  I hink this might be a race condition in some versions of postgres, but I'm not sure...
-  // the docs here: https://www.postgresql.org/docs/9.6/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
-  // say "Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages:
-  // CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended."
-  maybeSync(connection) {
-    if (this.isPreparedStatement) {
-      this._hasSentSync = true
-      connection.sync()
-    }
   }
 
   submit(connection) {
@@ -184,9 +173,12 @@ class Query extends EventEmitter {
       portal: this.portal,
       rows: rows,
     })
+    // if we're not reading pages of rows send the sync command
+    // to indicate the pipeline is finished
     if (!rows) {
-      this.maybeSync(connection)
+      connection.sync()
     } else {
+      // otherwise flush the call out to read more rows
       connection.flush()
     }
   }

--- a/packages/pg/lib/query.js
+++ b/packages/pg/lib/query.js
@@ -31,6 +31,7 @@ class Query extends EventEmitter {
     this.isPreparedStatement = false
     this._canceledDueToError = false
     this._promise = null
+    this._hasSentSync = false
   }
 
   requiresPreparation() {
@@ -100,7 +101,8 @@ class Query extends EventEmitter {
     this._checkForMultirow()
     this._result.addCommandComplete(msg)
     // need to sync after each command complete of a prepared statement
-    if (this.isPreparedStatement) {
+    if (this.isPreparedStatement && !this._hasSentSync) {
+      this._hasSentSync = true
       con.sync()
     }
   }
@@ -109,7 +111,8 @@ class Query extends EventEmitter {
   // the backend will send an emptyQuery message but *not* a command complete message
   // execution on the connection will hang until the backend receives a sync message
   handleEmptyQuery(con) {
-    if (this.isPreparedStatement) {
+    if (this.isPreparedStatement && !this._hasSentSync) {
+      this._hasSentSync = true
       con.sync()
     }
   }
@@ -126,7 +129,13 @@ class Query extends EventEmitter {
 
   handleError(err, connection) {
     // need to sync after error during a prepared statement
-    if (this.isPreparedStatement) {
+    // in postgres 9.6 the backend sends both a command complete and error response
+    // to a query which has timed out on rare, random occasions.  If we send sync twice we will receive
+    // to 'readyForQuery' events.  I think this might be a bug in postgres 9.6, but I'm not sure...
+    // the docs here: https://www.postgresql.org/docs/9.6/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+    // say "Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages: CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended."
+    if (this.isPreparedStatement && !this._hasSentSync) {
+      this._hasSentSync = true
       connection.sync()
     }
     if (this._canceledDueToError) {

--- a/packages/pg/lib/query.js
+++ b/packages/pg/lib/query.js
@@ -136,11 +136,12 @@ class Query extends EventEmitter {
     this.emit('end', this._results)
   }
 
-  // in postgres 9.6 the backend sends both a command complete and error response
-  // to a query which has timed out on rare, random occasions.  If we send sync twice we will receive
-  // to 'readyForQuery' events.  I think this might be a bug in postgres 9.6, but I'm not sure...
+  // In postgres 9.x & 10.x the backend sends both a CommandComplete and ErrorResponse
+  // to the same query when it times out due to a statement_timeout on rare, random occasions.  If we send sync twice we will receive
+  // to ReadyForQuery messages .  I hink this might be a race condition in some versions of postgres, but I'm not sure...
   // the docs here: https://www.postgresql.org/docs/9.6/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
-  // say "Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages: CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended."
+  // say "Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages:
+  // CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended."
   maybeSync(connection) {
     if (this.isPreparedStatement && !this._hasSentSync) {
       this._hasSentSync = true

--- a/packages/pg/lib/query.js
+++ b/packages/pg/lib/query.js
@@ -97,24 +97,33 @@ class Query extends EventEmitter {
     }
   }
 
-  handleCommandComplete(msg, con) {
+  handleCommandComplete(msg, connection) {
     this._checkForMultirow()
     this._result.addCommandComplete(msg)
     // need to sync after each command complete of a prepared statement
-    if (this.isPreparedStatement && !this._hasSentSync) {
-      this._hasSentSync = true
-      con.sync()
-    }
+    this.maybeSync(connection)
   }
 
   // if a named prepared statement is created with empty query text
   // the backend will send an emptyQuery message but *not* a command complete message
   // execution on the connection will hang until the backend receives a sync message
-  handleEmptyQuery(con) {
-    if (this.isPreparedStatement && !this._hasSentSync) {
-      this._hasSentSync = true
-      con.sync()
+  handleEmptyQuery(connection) {
+    this.maybeSync(connection)
+  }
+
+  handleError(err, connection) {
+    // need to sync after error during a prepared statement
+    this.maybeSync(connection)
+    if (this._canceledDueToError) {
+      err = this._canceledDueToError
+      this._canceledDueToError = false
     }
+    // if callback supplied do not emit error event as uncaught error
+    // events will bubble up to node process
+    if (this.callback) {
+      return this.callback(err)
+    }
+    this.emit('error', err)
   }
 
   handleReadyForQuery(con) {
@@ -127,27 +136,16 @@ class Query extends EventEmitter {
     this.emit('end', this._results)
   }
 
-  handleError(err, connection) {
-    // need to sync after error during a prepared statement
-    // in postgres 9.6 the backend sends both a command complete and error response
-    // to a query which has timed out on rare, random occasions.  If we send sync twice we will receive
-    // to 'readyForQuery' events.  I think this might be a bug in postgres 9.6, but I'm not sure...
-    // the docs here: https://www.postgresql.org/docs/9.6/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
-    // say "Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages: CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended."
-    if (this.isPreparedStatement && !this._hasSentSync) {
+  // in postgres 9.6 the backend sends both a command complete and error response
+  // to a query which has timed out on rare, random occasions.  If we send sync twice we will receive
+  // to 'readyForQuery' events.  I think this might be a bug in postgres 9.6, but I'm not sure...
+  // the docs here: https://www.postgresql.org/docs/9.6/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+  // say "Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages: CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended."
+  maybeSync(connection) {
+    if (this.isPreparedStatement) {
       this._hasSentSync = true
       connection.sync()
     }
-    if (this._canceledDueToError) {
-      err = this._canceledDueToError
-      this._canceledDueToError = false
-    }
-    // if callback supplied do not emit error event as uncaught error
-    // events will bubble up to node process
-    if (this.callback) {
-      return this.callback(err)
-    }
-    this.emit('error', err)
   }
 
   submit(connection) {

--- a/packages/pg/lib/query.js
+++ b/packages/pg/lib/query.js
@@ -142,7 +142,7 @@ class Query extends EventEmitter {
   // the docs here: https://www.postgresql.org/docs/9.6/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
   // say "Therefore, an Execute phase is always terminated by the appearance of exactly one of these messages: CommandComplete, EmptyQueryResponse (if the portal was created from an empty query string), ErrorResponse, or PortalSuspended."
   maybeSync(connection) {
-    if (this.isPreparedStatement) {
+    if (this.isPreparedStatement && !this._hasSentSync) {
       this._hasSentSync = true
       connection.sync()
     }

--- a/packages/pg/test/integration/client/prepared-statement-tests.js
+++ b/packages/pg/test/integration/client/prepared-statement-tests.js
@@ -174,5 +174,15 @@ var suite = new helper.Suite()
     checkForResults(query)
   })
 
+  suite.testAsync('with no data response and rows', async function () {
+    const result = await client.query({
+      name: 'some insert',
+      text: '',
+      values: [],
+      rows: 1,
+    })
+    assert.equal(result.rows.length, 0)
+  })
+
   suite.test('cleanup', () => client.end())
 })()

--- a/packages/pg/test/integration/gh-issues/1105-tests.js
+++ b/packages/pg/test/integration/gh-issues/1105-tests.js
@@ -6,7 +6,7 @@ suite.testAsync('timeout causing query crashes', async () => {
   const client = new helper.Client()
   await client.connect()
   await client.query('CREATE TEMP TABLE foobar( name TEXT NOT NULL, id SERIAL)')
-  client.query('BEGIN')
+  await client.query('BEGIN')
   await client.query("SET LOCAL statement_timeout TO '1ms'")
   let count = 0
   while (count++ < 5000) {

--- a/packages/pg/test/integration/gh-issues/1105-tests.js
+++ b/packages/pg/test/integration/gh-issues/1105-tests.js
@@ -1,0 +1,20 @@
+const pg = require('../../../lib')
+const helper = require('../test-helper')
+const suite = new helper.Suite()
+
+suite.testAsync('timeout causing query crashes', async () => {
+  const client = new helper.Client()
+  await client.connect()
+  await client.query('CREATE TEMP TABLE foobar( name TEXT NOT NULL, id SERIAL)')
+  client.query('BEGIN')
+  await client.query("SET LOCAL statement_timeout TO '1ms'")
+  let count = 0
+  while (count++ < 5000) {
+    try {
+      await client.query('INSERT INTO foobar(name) VALUES ($1)', [Math.random() * 1000 + ''])
+    } catch (e) {
+      await client.query('ROLLBACK')
+    }
+  }
+  await client.end()
+})

--- a/packages/pg/test/integration/gh-issues/2085-tests.js
+++ b/packages/pg/test/integration/gh-issues/2085-tests.js
@@ -4,6 +4,10 @@ var assert = require('assert')
 
 const suite = new helper.Suite()
 
+if (process.env.PGTESTNOSSL) {
+  return
+}
+
 suite.testAsync('it should connect over ssl', async () => {
   const ssl = helper.args.native
     ? 'require'

--- a/packages/pg/test/integration/gh-issues/2085-tests.js
+++ b/packages/pg/test/integration/gh-issues/2085-tests.js
@@ -4,6 +4,8 @@ var assert = require('assert')
 
 const suite = new helper.Suite()
 
+// allow skipping of this test via env var for
+// local testing when you don't have SSL set up
 if (process.env.PGTESTNOSSL) {
   return
 }

